The request System. Or as it is commonly known the DMan 2.0

marchermans opened the pull request:
This is a discussion PR.

Ist main goal is to get Input on the design of the new System.
As you all can see it is heavily Interface and factory based.

The principle follows an Inversion of Control pattern that makes the IRequestManager the heart of the System. Outside of him no IRequests are known, just Tokens and These Tokens are passed around.

If you request a IRequest from the Manager you will get a defensive copy to make sure that you cannot tamper with its state directly.

RequestResolvers are the workers in this System. A resolver basicly tells the Handler what to do. And how to proceed.

Please feel free to ask me anything regarding the new System.

marchermans committed:
Started working on the request system.
Core interfaces defined yet not finalized.

Implementations provided for an ItemStack request and result.

A resolver should be implemented on a case to case basis, eg a WarehouseResolver should function different from a SmithsResolver.

Signed-off-by: OrionOnline <oriondevelopment@outlook.com>

Commit:
Working on creating a proper way to work with ChainedRequests and construction, serializing and deserializing requests.

Added:
   * IRequestFactory, will in the handle the creation, serialization and deserialization of Request and there relevant Data.

 Updated ChainedRequest:
    * Still trying to figure out on how to implement this properly in a generic way.
    * First attempt to get this working.

marchermans committed:
Started working on chaining requests, also started working on construction, serializing and deserializing Requests.

Added a definition for requestfactories.
Started working on a implementation of Chained Requests, i am not yet happy with those however.

Signed-off-by: OrionOnline <oriondevelopment@outlook.com>

marchermans committed:
Merge remote-tracking branch 'origin/feature/api-rework' into feature/api-rework

# Conflicts:
#	src/main/java/com/minecolonies/coremod/colony/management/requestsystem/api/IRequestFactory.java

marchermans committed:
Continued working on requests, refactored some original code. Created tokens to represent requests outside of the system.

Added:
  * Request tokens. Represent requests in a colony outside of the management system
    That way a Worker can hold a list of token of requests he as made and have been made to him,
    without having an instance that he has to hold and the burden that comes with that (serialization, syncing etc)
  * StandardRequestToken. Implements IRequestToken, probably only class that will be needed, however it should be noted
    that the interface is still provided for compatibility in the future. Also note that its equals() and hashCode()
    methods are overloaded to allow it to be used in Maps as a key without requiring instance unicy between workers and
    managers, just ID unicy.
  * AbstractRequest. An abstract class that holds a basic implementation of a request.

Started:
  * Manager definition is being created.

Changed:
  * Resolving. An additional resolving method was added: attemptResolve(IRequest request) returns a list of requirements
    that need to be completed before resolve(IRequest request) can be called. Is the returned list empty no requirements
    are needed. Is it null the resolve failed, and the next resolver will be tried.

Signed-off-by: OrionOnline <oriondevelopment@outlook.com>

marchermans committed:
Finished the definition of the api of the request system.

The definition is with reservation!

Added:
  * ILocatable: Interface to get mark an object as locatable in the real world.
                When creating a request a locatable has to be provided as the requester.
  * ILocation: An interface describing a location in the Minecraft-universe.
               It combines a BlockPos and a Dimension together.
               It can be both implemented as a dynamic system or as a static.
  * ILocationFactory: Interface describing a class that can serialize and deserialize ILocation objects
                      from and to NBT.
  * EntityLocation: An ILocation implementation that references Entities. Also provides a factory as innerclass.
  * StaticLocation: An ILocation implementation that references static locations. Also provides a factory as innerclass.
  * Delivery: A requestable that describes a delivery. Can be used to request deliveries just like ItemStacks.

Signed-off-by: OrionOnline <oriondevelopment@outlook.com>

marchermans committed:
Refactored the Factory system.

Added:
  * IFactory: Definition for a Factory class
  * IFactoryController: Definition for a class that manages factories.
  * StandardFactoryController: Default implementation for the IFactoryController interface.
  * IRequestTokenFactory: Factory for request tokens. Follows the IFactory definition.
  * StandardRequestFactories: Implementation for the default factories of the requests that are part of minecolonies.
  * StandardRequestTokenFactory: Factory for the StandardRequestToken.

Changed:
  * ILocationFactory / IRequestFactory: Updated to reflect the new Factory system. Also added a wrapper default implementation for the newInstance method.
  * IRequest: Updated to allow better handling inside the factory. Added some helper methods.
  * IRequestToken: ----Moved----
  * IRequestManager: Removed the factory retrieval methods. Added a method to get the FactoryController for the manager.
  * IRequestResolver: Added a method that is called when the user overrides a request by delivering it himself.
  * RequestState: Added additional states to allow better handling of states.
  * EntityLocation: Changed the entity reference, to a weak reference of type entity. Prevents a MemLeak.
  * StaticLocation: Updated the factory to reflect the changes to IFactory.
  * Delivery: Added serialization and deserialization helper methods.
  * AbstractRequest: Implemented the changes to reflect those from IRequest.
  * StandardRequests: Updated to reflect the changes in the AbstractRequest class. Overridden the hasResult method in the ItemStack variant to reflect ItemStack.EMPTY.
  * StandardRequestToken: ----Moved----

Signed-off-by: OrionOnline <oriondevelopment@outlook.com>

marchermans committed:
Much work.

Signed-off-by: OrionOnline <oriondevelopment@outlook.com>

marchermans committed:
Working on the manager implementation.

Signed-off-by: OrionOnline <oriondevelopment@outlook.com>

marchermans committed:
Working on the core mechanics some more.
Also added a bunch of logging.

Signed-off-by: OrionOnline <oriondevelopment@outlook.com>

marchermans committed:
Worked on a bunch of core implementation stuff.

Updated some api interfaces to make some methods clearer and better defined.

Signed-off-by: OrionOnline <oriondevelopment@outlook.com>

marchermans committed:
Finished Core implementation.

Added a way to use the player as a fallback.

TODO:
  * Implement resolvers. Including the player one.

Signed-off-by: OrionOnline <oriondevelopment@outlook.com>

marchermans committed:
Initial work of the integration of the new system.

Signed-off-by: OrionOnline <oriondevelopment@outlook.com>

marchermans committed:
Initial work of the integration of the new system.

Signed-off-by: OrionOnline <oriondevelopment@outlook.com>

marchermans committed:
First working implementation of the Request system

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/Minecolonies/minecolonies?pullRequest=1001) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you all sign our [Contributor License Agreement](https://cla-assistant.io/Minecolonies/minecolonies?pullRequest=1001) before we can accept your contribution.<br/>**2** out of **3** committers have signed the CLA.<br/><br/>:white_check_mark: OrionDevelopment<br/>:white_check_mark: Kostronor<br/>:x: Marc Hermans<br/><hr/>**Marc Hermans** seems not to be a GitHub user. You need a GitHub account to be able to sign the CLA. If you have already a GitHub account, please [add the email address used for this commit to your account](https://help.github.com/articles/why-are-my-commits-linked-to-the-wrong-user/#commits-are-not-linked-to-any-user).<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/Minecolonies/minecolonies?pullRequest=1001) it.</sub>

marchermans commented:
Yes there are a bunch of Things that i Need to merge. This pr is just for discussion and should be treated as such.

marvin-bitterlich commented on src/api/java/com/minecolonies/coremod/colony/permissions/Action.java:
```diff
@@ -0,0 +1,36 @@
+    /**
+     * Stores the action as byte.
+     * {@link #ACCESS_HUTS} has value 0000 0000
+     * {@link #SEND_MESSAGES} has value 0100 0000
```
showing only 8 of 32 bits makes it seem like a bug ^^

marvin-bitterlich commented on src/api/java/com/minecolonies/coremod/colony/requestsystem/RequestState.java:
```diff
@@ -0,0 +1,86 @@
+         * This should never be changed! It is used to read and write from NBT so it has to
+         * persist between mod versions.
+         */
+        for(RequestState state : RequestState.values())
```
optional braces?

Raycoms commented on src/api/java/com/minecolonies/coremod/colony/requestsystem/IRequestManager.java:
```diff
@@ -0,0 +1,95 @@
+     * @param object The Object that is being requested.
+     * @param <T> The type of request.
+     * @return The token representing the request.
+     * @throws IllegalArgumentException is thrown when this manager cannot produce a request for the given types.
```
Should we really throw illegalArgument exceptions or should we return something like Token.INVALID?

marvin-bitterlich replied:
exeption is better style then null value imo

Raycoms replied:
I wouldn't make it a null value. I would make it really a Token.Invalid with a concrete value.
We had a lot of exceptions in minecolonies which we removed almost completely because they did more bad than good.

marchermans replied:
I would like to see which type you would use?

Raycoms replied:
Its an IToken but it has certain values so it equals IToken.INVALID.

marchermans replied:
Itoken is an interface. Technically just hull around a different object. The standard uses a uuid. However i think this is better it is stated in the method header that this method throws the exception so everybody knows it does.

Raycoms commented on src/api/java/com/minecolonies/coremod/colony/requestsystem/RequestState.java:
```diff
@@ -0,0 +1,86 @@
+    /**
+     * State for a request when it has been forcefully fullfilled by a player
+     */
+    OVERRULED,
```
Nice, the states are great now.

Raycoms replied:
The formatting of this file is off

marvin-bitterlich replied:
formatting is off everywhere, @OrionDevelopment will fix it prior to merge

marvin-bitterlich commented on src/main/java/com/minecolonies/coremod/colony/Colony.java:
```diff
@@ -1436,8 +1440,19 @@ public void addWayPoint(final BlockPos point, IBlockState block)
      */
     @NotNull
-    public Map<BlockPos, AbstractBuilding> getBuildings()
+    public ImmutableMap<BlockPos, IBuilding> getBuildings()
```
does it have a reason you return Immuteablemap?

marchermans replied:
Why would it not Need to be an Immutable. Do you want something from outside your colony to mess up the map, ofcourse not. If it would not be an immutable, somebody might remove a Building and the request System does not get notified.

Raycoms replied:
There is also the option of "Unmodifiable map"

Raycoms commented on src/api/java/com/minecolonies/coremod/colony/requestsystem/location/ILocation.java:
```diff
@@ -0,0 +1,31 @@
+
+/**
+ * Interface used to describe locations in the world.
+ */
```
Colonist are currently unable to switch dimensions, should take care with that, it causes a lot of problems with them atm.

marvin-bitterlich replied:
still fine to record what dim we are in

Raycoms replied:
Just noted we need to take care of that =D

marvin-bitterlich commented on src/main/java/com/minecolonies/coremod/colony/jobs/AbstractJob.java:
```diff
@@ -154,12 +150,7 @@ public static AbstractJob createFromNBT(final CitizenData citizen, @NotNull fina
-            final NBTTagCompound itemCompound = itemsNeededTag.getCompoundTagAt(i);
-            itemsNeeded.add(new ItemStack(itemCompound));
-        }
+        //NOOP; Requests are stored on the building.
```
Nice 😍

marchermans replied:
:P It was one of the Corner Stones. Request are state based, so hey are stored in something else.

marvin-bitterlich commented on src/main/java/com/minecolonies/coremod/colony/permissions/Permissions.java:
```diff
@@ -714,73 +714,6 @@ public void serializeViewNetworkData(@NotNull final ByteBuf buf, @NotNull final
     /**
-     * Ranks within a colony.
-     */
-    public enum Rank
```
👍

marvin-bitterlich commented on src/main/java/com/minecolonies/coremod/colony/requestsystem/StandardFactoryController.java:
```diff
@@ -0,0 +1,156 @@
+ */
+public class StandardFactoryController implements IFactoryController {
+
+    ////// --------------------------- NBTConstants --------------------------- \\\\\\
```
Cool 👓

marvin-bitterlich commented on src/main/java/com/minecolonies/coremod/entity/ai/basic/AbstractEntityAIStructure.java:
```diff
@@ -1078,13 +1079,54 @@ else if (block instanceof BlockFlowerPot)
-            building.addContainerPosition(pos);
+
+        try{
+            if (job instanceof JobBuilder) {
```
can we make this go away? ^^

xavierh replied:
Its in my pr already I think!

Raycoms replied:
That should be away already I think. I rewrote all that builder stuff already.
Probably just has to merge the newest changes to his branch

marvin-bitterlich commented on src/main/resources/assets/minecolonies/lang/en_us.lang:
```diff
@@ -215,6 +217,3 @@ tile.minecolonies.blockHutWareHouse.name=Warehouse
 tile.minecolonies.blockHutDeliveryman.name=Deliveryman
 com.minecolonies.coremod.gui.workerHuts.Deliveryman=Deliveryman
-com.minecolonies.coremod.gui.workerHuts.cancelBuild=Cancel Build
-com.minecolonies.coremod.gui.workerHuts.cancelUpgrade=Cancel Upgrade
```
why?

Raycoms replied:
Probably because old prs are not merged yet.

Raycoms commented on src/main/java/com/minecolonies/coremod/colony/requestsystem/requestable/Tool.java:
```diff
@@ -0,0 +1,69 @@
+import org.jetbrains.annotations.NotNull;
+
+/**
+ * Class used to represent tools inside the request system.
```
Nice, we really needed that

marvin-bitterlich replied:
🎉

xavierh commented on src/api/java/com/minecolonies/coremod/colony/IColony.java:
```diff
@@ -61,4 +66,27 @@
+     * @return The buildings in the colony
+     */
+    @NotNull
+    ImmutableMap<BlockPos, IBuilding> getBuildings();
```
I think this belong to a different pr, and could be merge well before dman 2.0 is ready

xavierh commented on src/api/java/com/minecolonies/coremod/colony/buildings/IBuilding.java:
```diff
@@ -0,0 +1,53 @@
+/**
+ * Created by marcf on 3/7/2017.
+ */
+public interface IBuilding {
```
I think this belong to a different pr, and could be merge well before dman 2.0 is ready

xavierh commented on src/api/java/com/minecolonies/coremod/colony/permissions/Action.java:
```diff
@@ -0,0 +1,36 @@
+package com.minecolonies.coremod.colony.permissions;
```
I think this belong to a different pr, and could be merge well before dman 2.0 is ready

xavierh commented on src/api/java/com/minecolonies/coremod/colony/permissions/IPermissions.java:
```diff
@@ -14,10 +14,10 @@
      * @return true if has permission, otherwise false.
      */
-    boolean hasPermission(EntityPlayer player, Permissions.Action action);
+    boolean hasPermission(EntityPlayer player, Action action);
```
I think this belong to a different pr, and could be merge well before dman 2.0 is ready

xavierh commented on src/api/java/com/minecolonies/coremod/colony/permissions/Rank.java:
```diff
@@ -0,0 +1,33 @@
+package com.minecolonies.coremod.colony.permissions;
```
I think this belong to a different pr, and could be merge well before dman 2.0 is ready

xavierh commented on src/api/java/com/minecolonies/coremod/colony/requestsystem/location/ILocatable.java:
```diff
@@ -0,0 +1,16 @@
+package com.minecolonies.coremod.colony.requestsystem.location;
```
I think this belong to a different pr, and could be merge well before dman 2.0 is ready

xavierh commented on src/api/java/com/minecolonies/coremod/colony/requestsystem/location/ILocation.java:
```diff
@@ -0,0 +1,31 @@
+package com.minecolonies.coremod.colony.requestsystem.location;
```
I think this belong to a different pr, and could be merge well before dman 2.0 is ready

xavierh commented on src/main/java/com/minecolonies/coremod/blocks/AbstractBlockHut.java:
```diff
@@ -202,7 +202,7 @@ public boolean onBlockActivated(
             if (building != null
                   && building.getColony() != null
-                  && building.getColony().getPermissions().hasPermission(playerIn, Permissions.Action.ACCESS_HUTS))
+                  && building.getColony().getPermissions().hasPermission(playerIn, Action.ACCESS_HUTS))
```
I think this belong to a different pr, and could be merge well before dman 2.0 is ready

xavierh commented on src/main/java/com/minecolonies/coremod/client/gui/WindowTownHall.java:
```diff
@@ -302,7 +303,7 @@ private void updateUsers()
         users.clear();
         users.addAll(townHall.getColony().getPlayers().values());
-        users.sort(Comparator.comparing(Permissions.Player::getRank, Permissions.Rank::compareTo));
+        users.sort(Comparator.comparing(Permissions.Player::getRank, Rank::compareTo));
```
I think this belong to a different pr, and could be merge well before dman 2.0 is ready

xavierh commented on src/main/java/com/minecolonies/coremod/client/gui/WindowTownHall.java:
```diff
@@ -653,7 +654,7 @@ private void removePlayerClicked(final Button button)
         {
             final Permissions.Player user = users.get(row);
-            if (user.getRank() != Permissions.Rank.OWNER)
+            if (user.getRank() != Rank.OWNER)
```
I think this belong to a different pr, and could be merge well before dman 2.0 is ready

xavierh commented on src/main/java/com/minecolonies/coremod/colony/CitizenData.java:
```diff
@@ -20,7 +23,7 @@
  * Extra data for Citizens.
  */
-public class CitizenData
+public class CitizenData implements ILocatable
```
I think this belong to a different pr, and could be merge well before dman 2.0 is ready

xavierh reviewed (changes requested):
A lost of (good) stuff should be in a different PR and could be added well before the Request system is done
My old brain can not take in +4000 lines in one go.
some interface should be add earlier so that other devs can star using them.

marchermans committed:
Working on Merging stuff from version/1.11.

marchermans committed:
Merge remote-tracking branch 'origin/testing/1.11' into Feature/api-rework

marchermans committed:
Merged Testing/1.11.

marchermans committed:
Working again on a bunch of changes related to the API.

cltnschlosser commented:
I agree with @xavierh, there is way too much going on in this pull request. We need take a step back and get some info written down about what is going on here, and then create sub tasks with smaller goals. I'm sure there is a lot of valuable code here, but it's way to hard to tell. Let's get a plan together and start merging this piece by piece. It's in everyones best interest to not have long lived branches with tons of changes on them. I'd hate to see all this work get so separated from the mod that it can't be merged in.

This pull request also appears to be trying to separate the mod into an api (set of interfaces) and an implementation, but also implementing a request system. I think those should be done in separate pull requests.

Also, I think that if we are going to pull out an api, that instead of making it a separate package, we should make it a separate module, that our implementation depends on.

I would like to start moving some of our Enums and static maps to registries, which would be beneficial for api work, but it's hard for me to tell what has been done.

marchermans commented:
So lets start from the beginning.
This Pull is an API, from scratch designed.
It will consist out of a Major rework and an even bigger refactoring.

# What does it contain when i am done:
## Module wise:
- A _Module_ for the API. It is needed to create an API Jar.
- A _Module_ for BlockOut. It is not needed, but creates a more cohesive package structure. Also would allow for better Distribution if we ever needed it again.
- A _Module_ for Structures. (Same as with the BlockOut Module)
## API wise:
- Interfaces and or Abstract classes for most of our Core classes, which include but are not limited too:
    - Colony and ColonyView -> IColony
    - AbstractBuilding And View -> IBuilding
    - AbstractJob -> IJob
    - etc.
- Utility Classes and the complete Lib-Package
- The Request System (at the heart of this PR, which is what it originally started as)
- A factory Pattern that should be used to create, serialize and deserialize new Objects that are managed by the API.
- A new ID System for Buildings (alltough rightnow i redirected all the calls to get ID() which returned a BlockPos and now Returns an IToken<UUID> to getLocation().getInDimensionLocation(), as getLocation() now Returns an ILocation, to be specific it returns a StaticLocation, hence the getInDimensionLocation Call). This new System is needed to identify Buildings properly inside the Request System, as from its standpoint the Building functions as the requester of a given request, even though it just manages the request for ist citizens.
## Refactoring wise:
- Due to the Need of the IColony Interface being in the API package, this caused a cascading downfall of classes either needing to be moved, renamed, replaced or extended somehow. A lot of refactoring also went into the readability of the Code. A lot of classes are entangled with others, without having an Interface in between to handle the interaction in proper OOP style. Refactoring this so i could use them properly is the biggest part of this PR in the first place.

# What is not in this PR:
- Registries for custom Objects. As of right now this is not part of this PR, because it is not directly needed for the Request System to function properly.
- Abstraction layers to make the API more stable. Yes these Need to be introduced.

# Why is this not done in small Pull Request.
- Well first of all, have you ever tried pulling IColony out the main Module and into an other one?. The amount of compile Errors it creates due to classes missing is enormous. Due to the lack of abstraction between the Interface and say AbstractBuilding (and all that that class depends upon) it is not possible to create a compiling Version of the Code. From the Moment I was done with the core API of the request System i have been working on getting it to compile, but i am still rewriting stuff to make the move of the IColony Interface work.
- Yes i Agree that their are some **small** changes that could have been (and maybe should be) in a different PR. However I am not a Person that commits after every file i have changed, nor should i be forced to when i make changes of this size. So i would like to see somebody cherrypick the files that he wants. If he can do it and make a PR he is more then welcome to do so.

TLDR: This is still not compiling because i moved IColony and it resulting in a Compiler outcry. And that is the reason why this is a big one push PR and not twenty different small ones.

cltnschlosser commented:
Well cool, I'll start on the registries then. But even from looking at your bullet points I see 4-5 PRs

marchermans commented:
I would like to ask you to wait with the registries, atleast untill i have this compiling. We can then discuss in how many PRS i Need to split this.

marchermans committed:
Further API Work.

Added:
  * Modules for BlockOut
  * Modules for Structures

marchermans committed:
Further improvements on the refactoring.

ColonyManager is last but not least.
Added IAPI, IAPI.Holder and API.

marchermans committed:
Fixing some small gradle stuff.

marchermans committed:
Fix for using the wrong logger in the API.

marvin-bitterlich committed:
Changing which modules depend on the API, plus some refactorings.

marchermans committed:
Working on the ColonyManager.

marchermans committed:
Start working on IWorldColonyControllers.

marchermans committed:
Initial working version of the ColonyManager.

marchermans committed:
Setting the IWorldColonyController on the SaveData when a new instance is created.

marchermans committed:
Formatting following the Minecolonies Formatting guidelines for IDEA

marchermans committed:
Introducting the Legacy Colony Manager Loader.

marchermans committed:
Attaching the Legacy Loader to the existing system.

marchermans committed:
Enabling delayed loading of the WorldSaveData and added documentation.

marchermans committed:
Working on moving a bunch of Util classes again.

marchermans committed:
Working some more on the API.
